src/storage/txn/actions/check_txn_status.rs

```diff
@@ -5,16 +5,118 @@ use tikv_kv::SnapshotExt;
+            resolving_pessimistic_lock,
+        ) {
+            Ok(txn_status) => match txn_status {
+                t @ TxnStatus::RolledBack | t @ TxnStatus::Committed { .. } => (t, true),
+                _ => {
+                    return Err(ErrorInner::Other(box_err!(
+                        "unexpected result={:?} checking persist information \
+                            for pessimistic lock with conflict lock={:?}",
+                        &txn_status,
+                        lock
+                    ))
+                    .into());
+                }
+            },
+            Err(err) => match err {
+                // Continue to process if there's no corresponding persistent information.
+                // In this case the pessimistic primary lock should not be stale lock with conflict.
+                mvcc::Error(box ErrorInner::TxnNotFound { .. }) => {
+                    (TxnStatus::LockNotExistDoNothing, false)
+                }
+                _ => {
+                    return Err(err);
+                }
+            },
+        };
+        if is_current_lock_stale {
+            info!("unlock stale pessimistic primary lock";
+                "primary_key" => ?&primary_key,
+                "lock" => ?&lock,
+                "current_ts" => current_ts,
+                "resolving_pessimistic_lock" => ?resolving_pessimistic_lock,
+            );
+            let released = txn.unlock_key(primary_key, true, TimeStamp::zero());
+            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.pessimistic_rollback.inc();
+            return Ok((txn_status, released));
+        }
+    }
+
+    if lock.ts.physical() + lock.ttl < current_ts.physical() {
+        // The primary pessimistic lock has expired.
+        // If the resolving lock is a prewrite lock, the transaction must already be in
+        // the commit phase, the primary lock would no longer change in this
+        // case.
+        return if resolving_pessimistic_lock {
+            let released = txn.unlock_key(primary_key, true, TimeStamp::zero());
+            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.pessimistic_rollback.inc();
+            Ok((TxnStatus::PessimisticRollBack, released))
+        } else {
+            let released = rollback_lock(txn, reader, primary_key, lock, true, true)?;
+            MVCC_CHECK_TXN_STATUS_COUNTER_VEC.rollback.inc();
+            Ok((TxnStatus::TtlExpire, released))
+        };
+    }
+
+    // Just use LockNotExistDoNothing to avoid clone the lock, it would not be used.
+    Ok((TxnStatus::LockNotExistDoNothing, None))
+}
+
+/// Evaluate transaction status if a lock exists with the anticipated
+/// 'start_ts'.
+///
+/// 1. Validate whether the existing lock indeed corresponds to the
+/// primary lock. The primary key    may switch under certain circumstances. If
+/// it's a stale lock, the transaction status should    not be determined by it.
+/// Refer to https://github.com/tikv/tikv/issues/14636 for additional information.
```

Comment:
```suggestion
/// Refer to https://github.com/pingcap/tidb/issues/42937 for additional information.
```

#14636 doesn't say anything in detail. I think linking to the corresponding issue in TiDB repo is better.